### PR TITLE
Separate statements from GenTree, part 2

### PR DESCRIPTION
**Pull request by sandreenko**
Contributes to #25899.

All comparing are with the baseline without #25898 (because it added separated IL_OFFSET nodes to make this change simpler) and for x64 Release without PGO:

asm-diffs -c/-f crossge/pmi: no diffs

PIN System.Private.Corelib (10 runs): 0.08% improvement 🤣 

JitMemStats: 1.7% improvement:

|            | base                                | diff                                |
|------------|-------------------------------------|-------------------------------------|
| alloc size | 2204212637 (avg   71981 per method) | 2167454293 (avg   70780 per method) |
| ASTNode    | 431130000,  19.56%                  | 392179968,  18.09%                  |
| BasicBlock | 93342880,   4.23%                   | 95534568,   4.41%                   |

BasicBlock increase can be eliminated (if we put `bbTreeList` and `bbStmtList` in the same union, but it is not easy, because `bbTreeList` lives in `BasicBlock` base class: `LIR::Range`), but can be done in theory.

Statement struct could be smaller if we eliminate `bool compilerAdded;` that is used only in `optimizer`, but that is left for the future PRs.

**Changes:**

f0824b4b92: Add unreached for GT_STMT in TryGetUse, GenTreeUseEdgeIterator, GetChild.

fca76286fd: Do not call gtDispTree for statements.

Call it for gtStmtExpr instead. We can rename/change it later, probably add a debug int for statement number.

ca192abca9: Separate bbTreeList and bbStmtList.

bbTreeList is used for LIR, bbStmtList for the previous phases.
They can't be both set at the same time.

54eab27dc6: Fix rationalize.

Because now bbStmtList and bbTreeList, GT_STMT and GT_IL_OFFSET are separated
we can simplify logic when switch to LIR.

db660dd654: Create gtDispStmtList.

To make it obvious that we expect statements.

7637e94b9b: Fix another forgotten case of tree instead of stmt.

78dadce8de: Separate Stmt and Tree.

Do not inherit GenTreeStmt from GenTree.

af38991d0b: Update GetIndirSize to handle no user case.

This doesn't change behavior for the Stmt case because a user that was GT_STMT was not able to be GT_ASG at the same time.

0023837ba4: BasicBlock::Contains uses LIR::Range iterator and can't be used for non-LIR case.

e949ca56ea: Delete leftovers.

dcfa2b3854: Response review, part 1.

e5ea03598b: Response review, part 2.

### CODE REVIEW

File: src/jit/compiler.hpp [hunk 1]
```diff
@@ -859,6 +859,11 @@ const char* genES2str(BitVecTraits* traits, EXPSET_TP set);
 const char* refCntWtd2str(unsigned refCntWtd);
 #endif
 
+void* GenTreeStmt::operator new(size_t sz, class Compiler* compiler)
```
Comment (mikedn) [inline src/jit/compiler.hpp]: Why is this needed?

File: src/jit/gentree.h [hunk 2]
```diff
@@ -4932,7 +4930,7 @@ struct GenTreeStmt : public GenTree
         }
         else
         {
-            return gtNext->AsStmt();
+            return gtNext;
```
Comment (mikedn) [inline src/jit/gentree.h]: The whole function is now just `return gtNext;`

File: src/jit/gentree.h [hunk 3]
```diff
@@ -4944,34 +4942,40 @@ struct GenTreeStmt : public GenTree
...
-
-        // Set the statement to have the same costs as the top node of the tree.
-        // This is used long before costs have been assigned, so we need to copy
-        // the raw costs.
-        CopyRawCosts(expr);
     }
 
 #if DEBUGGABLE_GENTREE
-    GenTreeStmt() : GenTree(GT_STMT, TYP_VOID)
+    GenTreeStmt()
```
Comment (mikedn) [inline src/jit/gentree.h]: I don't think the `DEBUGGABLE_GENTREE` ctor is needed anymore, it's no longer a `GenTree`.

File: src/jit/gentree.cpp [hunk 4]
```diff
@@ -15399,13 +15312,9 @@ bool GenTree::IsPhiDefn()
     return res;
 }
 
-bool GenTree::IsPhiDefnStmt()
+bool GenTreeStmt::IsPhiDefnStmt()
```
Comment (mikedn) [inline src/jit/gentree.cpp]: Maybe just remove this? There aren't many uses and it's just `gtStmtExpr->IsPhiDefn()`.

File: src/jit/lsrabuild.cpp [hunk 5]
```diff
@@ -1718,7 +1718,7 @@ void LinearScan::buildPhysRegRecords()
 //
 BasicBlock* getNonEmptyBlock(BasicBlock* block)
 {
-    while (block != nullptr && block->bbTreeList == nullptr)
+    while (block != nullptr && block->bbTreeList == nullptr && block->bbStmtList == nullptr)
```
Comment (mikedn) [inline src/jit/lsrabuild.cpp]: This is in LIR so `bbStmtList` shouldn't be relevant anymore?

File: src/jit/compiler.h [hunk 6]
```diff
@@ -2825,7 +2825,7 @@ class Compiler
     int gtGetLclVarName(unsigned lclNum, char* buf, unsigned buf_remaining);
     char* gtGetLclVarName(unsigned lclNum);
     void gtDispLclVar(unsigned varNum, bool padForBiggestDisp = true);
-    void gtDispTreeList(GenTree* tree, IndentStack* indentStack = nullptr);
+    void gtDispStmtList(GenTreeStmt* tree, IndentStack* indentStack = nullptr);
```
Comment (mikedn) [inline src/jit/compiler.h]: Would be nice to rename `tree` to `stmt`

File: src/jit/gentree.cpp [hunk 7]
```diff
@@ -11381,11 +11296,11 @@ void Compiler::gtDispArgList(GenTreeCall* call, IndentStack* indentStack)
 // Assumptions:
 //    'tree' must be a GT_LIST node
 
-void Compiler::gtDispTreeList(GenTree* tree, IndentStack* indentStack /* = nullptr */)
+void Compiler::gtDispStmtList(GenTreeStmt* stmts, IndentStack* indentStack /* = nullptr */)
 {
-    for (/*--*/; tree != nullptr; tree = tree->gtNext)
+    for (GenTreeStmt* stmt = stmts; stmt != nullptr; stmt = stmt->gtNextStmt)
```
Comment (mikedn) [inline src/jit/gentree.cpp]: Use `getNextStmt()` instead of `gtNextStmt`.

File: src/jit/morph.cpp [hunk 8]
```diff
@@ -18116,19 +18116,19 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
...
-            m_compiler->gtDispTree(stmt);
+            m_compiler->gtDispTree(stmt->gtStmtExpr);
             m_stmtModified = false;
         }
 #endif // DEBUG
 
         WalkTree(&stmt->gtStmtExpr, nullptr);
 
-        // We could have somethinge like STMT(IND(ADDR(LCL_VAR))) so we need to escape
+        // We could have something like STMT(IND(ADDR(LCL_VAR))) so we need to escape
```
Comment (mikedn) [inline src/jit/morph.cpp]: Maybe change to "We could have a statement like IND(ADDR(LCL_VAR))" since STMT is no longer an oper.

File: src/jit/gentree.h (hunk 2 above)
Comment (sandreenko) [reply src/jit/gentree.h]: Yes, but I am planning to do it as a separate big renaming PR, where `GenTreeStatement` would be renamed to smth like `Statement`?

File: src/jit/block.cpp [hunk 9]
```diff
@@ -661,7 +661,7 @@ void BasicBlock::MakeLIR(GenTree* firstNode, GenTree* lastNode)
 bool BasicBlock::IsLIR()
 {
     const bool isLIR = (bbFlags & BBF_IS_LIR) != 0;
-    assert((bbTreeList == nullptr) || ((isLIR) == !bbTreeList->IsStatement()));
+    assert((isLIR && bbStmtList == nullptr) || (!isLIR && bbTreeList == nullptr));
```
Comment (mikedn) [inline src/jit/block.cpp]: Not sure if I understood the condition before and not sure if I understand it now. Doesn't this make all empty blocks appear as LIR?

File: src/jit/flowgraph.cpp [hunk 10]
```diff
@@ -542,27 +542,17 @@ bool Compiler::fgBlockContainsStatementBounded(BasicBlock*  block,
...
+//    Returns the new (potentially) statement.
 //
 // Notes:
 //    If 'stmt' is not already a statement, a new statement is created from it.
 //    We always insert phi statements at the beginning.
 //    In other cases, if there are any phi assignments and/or an assignment of
 //    the GT_CATCH_ARG, we insert after those.
 
-GenTreeStmt* Compiler::fgInsertStmtAtBeg(BasicBlock* block, GenTree* node)
+GenTreeStmt* Compiler::fgInsertStmtAtBeg(BasicBlock* block, GenTreeStmt* stmt)
```
Comment (mikedn) [inline src/jit/flowgraph.cpp]: Perhaps the return type should be changed to void? It's not very useful to return what was passed in as parameter.

**Comment by sandreenko** [comment]
PTAL @CarolEidt @dotnet/jit-contrib 

I recommend to review it per commit.

File: src/jit/gentree.cpp (hunk 4 above)
Comment (sandreenko) [reply src/jit/gentree.cpp]: I will remember it for the next change if you don't mind.

File: src/jit/gentree.cpp [hunk 11]
```diff
@@ -11381,11 +11296,11 @@ void Compiler::gtDispArgList(GenTreeCall* call, IndentStack* indentStack)
...
     {
-        gtDispTree(tree, indentStack);
+        gtDispTree(stmt->gtStmtExpr, indentStack);
```
Comment (mikedn) [inline src/jit/gentree.cpp]: Shouldn't we display IL offset somewhere?

File: src/jit/block.cpp (hunk 9 above)
Comment (sandreenko) [reply src/jit/block.cpp]: (isLir == true) => ((bbStmtList == nullptr) && (bbTreeList == any)),
(isLir == false) => ((bbTreeList == nullptr) && (bbStmtList  == any)),
so it means lir should not have `bbStmtList `, hir should not have `bbTreeList `.

File: src/jit/gentree.cpp (hunk 11 above)
Comment (sandreenko) [reply src/jit/gentree.cpp]: We need a separate function to print statement, probably debug numbers for them, but I was not sure about a name for this function (looks like it should start with gt anymore), so postponed it.

File: src/jit/morph.cpp (hunk 8 above)
Comment (sandreenko) [reply src/jit/morph.cpp]: Fixed

File: src/jit/gentree.cpp (hunk 7 above)
Comment (sandreenko) [reply src/jit/gentree.cpp]: Fixed

File: src/jit/compiler.h (hunk 6 above)
Comment (sandreenko) [reply src/jit/compiler.h]: Fixed

File: src/jit/gentree.h (hunk 3 above)
Comment (sandreenko) [reply src/jit/gentree.h]: Fixed

File: src/jit/compiler.hpp (hunk 1 above)
Comment (sandreenko) [reply src/jit/compiler.hpp]: Fixed

File: src/jit/lsrabuild.cpp (hunk 5 above)
Comment (sandreenko) [reply src/jit/lsrabuild.cpp]: Fixed

File: src/jit/gentree.cpp (hunk 11 above)
Comment (mikedn) [reply src/jit/gentree.cpp]: These `gt` & co. prefixes are rather horrible. I'd probably just make it `DumpStatement` and call it a day.

File: src/jit/flowgraph.cpp (hunk 10 above)
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: All functions `GenTreeStmt* fgInsertStmtSomewhere` return an argument as a result, will change it later in a separate PR, because for some of them we have users:
`afterStmt = fgInsertStmtAfter(block, afterStmt, newStmt);` 

It came from the time when these functions accepted trees and could return a stmt created to box that tree.

File: src/jit/gentree.cpp (hunk 11 above)
Comment (sandreenko) [reply src/jit/gentree.cpp]: VS doesn't work well with defines so, in general, these prefixes help to search/grep all occurrences. I do not like any of them, but do not see a nice solution.

Comment (mikedn) [reply src/jit/gentree.cpp]: > VS doesn't work well with defines so, in general, these prefixes help to search/grep all occurrences. I do not like any of them, but do not see a nice solution.

They also hurt writing & reading code (there's something truly horrible about having to write and read `gtGetOp1`) and that's something developers do more often than searching for uses of a particular definitions. And it's not unusual for this weird practice to fail to work well at solving exactly the problem is supposed to solve - finding uses. Because everything `GenTree` related is `gt` prefixed you're out of luck if you're trying to find `GenTreeLclFld::gtFieldSeq` uses and not `GenTreeIcon::gtFieldSeq` uses.

There's a perfectly good way to find all uses of a function or data member - just rename it and compile. If the name happens to be bad and you're doing tons of changes around that function/data member then it's even better - rename it for good and include it in the PR so that reviewers can easily see what parts of the JIT code are affected by the change.

The only case where prefixes have some appeal is the `Compiler` class with its zillion members. Of course, it shouldn't be like that but it's a ton of work to split it so in the meantime using C style prefixes is a reasonable workaround to preserve one's sanity when dealing with `Compiler`. Sadly, even this is messed up in the JIT as lots of functions use the `fg` prefix even then more specific prefixes (`ssa`, `vn` etc.) would have made more sense.

File: src/jit/compiler.h [hunk 12]
```diff
@@ -5238,13 +5238,17 @@ class Compiler
 #endif
 
 public:
-    GenTreeStmt* fgInsertStmtAtEnd(BasicBlock* block, GenTree* node);
+    GenTreeStmt* fgInsertStmtAtEnd(BasicBlock* block, GenTreeStmt* stmt);
+    GenTreeStmt* fgInsertTreeAtEnd(BasicBlock* block, GenTree* tree);
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.h]: Fixed

File: src/jit/flowgraph.cpp [hunk 13]
```diff
@@ -16962,7 +16940,7 @@ void Compiler::fgDebugCheckUpdate()
 
     /* We check for these conditions:
      * no unreachable blocks  -> no blocks have countOfInEdges() = 0
-     * no empty blocks        -> no blocks have bbTreeList = 0
+     * no empty blocks        -> no blocks have bbTreeList = 0 in LIR and bbStmtList = 0 before LIR
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/flowgraph.cpp]: Fixed, thank you.

File: src/jit/compiler.cpp [hunk 14]
```diff
@@ -4747,8 +4747,8 @@ void Compiler::compCompile(void** methodCodePtr, ULONG* methodCodeSize, JitFlags
         {
             for (GenTreeStmt* stmt = block->firstStmt(); stmt != nullptr; stmt = stmt->getNextStmt())
             {
-                compSizeEstimate += stmt->GetCostSz();
-                compCycleEstimate += stmt->GetCostEx();
+                compSizeEstimate += stmt->gtStmtExpr->GetCostSz();
+                compCycleEstimate += stmt->gtStmtExpr->GetCostEx();
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/jit/compiler.cpp]: Fixed.

**Comment by sandreenko** [comment]
I updated the PR with the suggested changes. 
Also, the root issue  #25899 was updated with the future changes suggested here.

Thanks for the review @CarolEidt, @mikedn.

File: src/jit/gentree.cpp (hunk 11 above)
Comment (sandreenko) [reply src/jit/gentree.cpp]: I'm not a fan of prefixes either and I understand your position.
However, I think it would be better to create a new issue from this comment (design discussion?). I think the main question is how to do such change (one massive PR that requires days of searching and renaming? do it incrementally somehow?), its cost and its benefits.

I do not think that it is productive not to use prefixes for the new methods/classes and keep them for the old.
